src/data/sign.lean

```diff
@@ -0,0 +1,196 @@
+@[simp] lemma pos_eq_one     : pos = 1  := rfl
+
+/-- The multiplication on `sign_type`. -/
+def mul : sign_type → sign_type → sign_type
+| neg neg  := pos
+| neg zero := zero
+| neg pos  := neg
+| zero _   := zero
+| pos h    := h
+
+instance : has_mul sign_type := ⟨mul⟩
+
+/-- The less-than relation on signs. -/
+inductive le : sign_type → sign_type → Prop
+| of_neg (a) : le neg a
+| zero       : le zero zero
+| of_pos (a) : le a pos
+
+instance : has_le sign_type := ⟨le⟩
+
+instance : decidable_rel le :=
+λ a b, by cases a; cases b; exact is_false (by rintro ⟨⟩) <|> exact is_true (by constructor)
+
+/- We can define a `field` instance on `sign_type`, but it's not mathematically sensible,
+so we only define the `comm_group_with_zero`. -/
+instance : comm_group_with_zero sign_type :=
+{ zero            := 0,
+  one             := 1,
+  mul             := (*),
+  inv             := id,
+  mul_zero        := λ a, by cases a; refl,
+  zero_mul        := λ a, by cases a; refl,
+  mul_one         := λ a, by cases a; refl,
+  one_mul         := λ a, by cases a; refl,
+  mul_inv_cancel  := λ a ha,  by cases a; trivial,
+  mul_comm        := λ a b,   by casesm* _; refl,
+  mul_assoc       := λ a b c, by casesm* _; refl,
+  exists_pair_ne  := ⟨0, 1,   by rintro ⟨⟩⟩,
+  inv_zero        := rfl }
+
+instance : linear_order sign_type :=
+{ le           := (≤),
+  le_refl      := λ a, by cases a; constructor,
+  le_total     := λ a b, by casesm* _; dec_trivial,
+  le_antisymm  := λ a b ha hb, by casesm* _; refl,
+  le_trans     := λ a b c hab hbc, by casesm* _; constructor,
+  decidable_le := le.decidable_rel }
+
+instance : has_distrib_neg sign_type :=
+{ neg_neg := λ x, by cases x; refl,
+  neg_mul := λ x y, by casesm* _; refl,
+  mul_neg := λ x y, by casesm* _; refl,
+..sign_type.has_neg }
+
+/-- `sign_type` is equivalent to `fin 3`. -/
+def fin3_equiv : sign_type ≃* fin 3 :=
+{ to_fun :=  λ a, a.rec_on 0 (-1) 1,
+  inv_fun := λ a, match a with
+    | ⟨0, h⟩   := 0
+    | ⟨1, h⟩   := 1
+    | ⟨2, h⟩   := -1
+    | ⟨n+3, h⟩ := (h.not_le le_add_self).elim
+  end,
+  left_inv :=  λ a, by cases a; refl,
+  right_inv := λ a, match a with
+    | ⟨0, h⟩   := rfl
+    | ⟨1, h⟩   := rfl
+    | ⟨2, h⟩   := rfl
+    | ⟨n+3, h⟩ := (h.not_le le_add_self).elim
+  end,
+  map_mul' := λ x y, by casesm* _; refl }
+
+/-- Turn a `sign_type` into zero, one, or minus one. This is a coercion instance, but note it is
+only a `has_coe_t` instance: see note [use has_coe_t]. -/
+def cast {α} [has_zero α] [has_one α] [has_neg α] : sign_type → α
+| zero :=  0
+| pos  :=  1
+| neg  := -1
+
+@[simp] lemma cast_zero    {α} [has_zero α] [has_one α] [has_neg α] : cast 0 = (0 : α) := rfl
+@[simp] lemma cast_one     {α} [has_zero α] [has_one α] [has_neg α] : cast 1 = (1 : α) := rfl
+@[simp] lemma cast_neg_one {α} [has_zero α] [has_one α] [has_neg α] : cast (-1) = (-1 : α) := rfl
+
+instance {α} [has_zero α] [has_one α] [has_neg α] : has_coe_t sign_type α := ⟨cast⟩
```

Comment:
I foresee many annoyances if we mix both `cast (sign x)` and `↑(sign x)` (for example `alg_hom.to_ring_hom` versus using the coercion). Could you declare one of them (preferably the coercion) as `simp`-normal form by adding a `@[simp] lemma cast_eq_coe` (or vice versa)?